### PR TITLE
fix: populate language on federated statuses so Translate control appears

### PR DESCRIPTION
**Pull request by llun**
Federated posts ingested over ActivityPub never had their language
populated, so the per-status Translate control (PR #925) never rendered
on them. Derive the ISO 639-1 language from the incoming Note's
contentMap/summaryMap locale key and persist it.

- Add getLanguage(note) helper in lib/activities/note.ts that reads the
  first contentMap key (falling back to summaryMap), normalized via the
  shared normalizeLanguageCode. Returns null for the array/Wordpress
  contentMap shape or when nothing is resolvable.
- Pass the derived language into database.createNote in createNoteJob;
  the announce/boost path reuses createNoteJob so it is covered too.
- Refresh language on federated edits in updateNoteJob, preserving the
  existing value when the edit carries no locale.
- Add a documented best-effort backfill migration: the stored content
  blob does not retain the original contentMap, so historical rows
  cannot be recovered without inventing data (intentional no-op).
- Unit-test getLanguage and assert createNoteJob stores language 'th'
  for an incoming Thai contentMap.

https://claude.ai/code/session_01F5vFHXSo1KtyxEoT7EPs5s

### CODE REVIEW

File: lib/activities/note.ts [hunk 1]
```diff
@@ -93,6 +94,29 @@ export const getContent = (object: BaseNote) => {
...
 }
 
+const firstLocaleKey = (
+  map: Record<string, string> | string[] | null | undefined
+): string | undefined => {
+  // Only locale-keyed objects encode a language; the array/Wordpress shape
+  // carries no locale information.
+  if (!map || Array.isArray(map)) return undefined
+  return Object.keys(map)[0]
+}
```
[the remark this answers is not in the document]
Comment (llun) [reply lib/activities/note.ts:105]: Good catch — added the defensive `typeof map !== 'object'` guard so a malformed AP payload where `contentMap`/`summaryMap` arrives as a string/number (or any non-object) is treated as "no locale" instead of risking an `Object.keys` surprise. Fixed in 4185af3.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: migrations/20260609101100_backfill_status_language.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+  // source language of historical federated statuses cannot be recovered.
+}
+
+/**
+ * @param { import("knex").Knex } knex
+ * @returns { Promise<void> }
+ */
+exports.down = async function () {
+  // Nothing to undo: the up migration does not modify any data.
+}
```
[the remark this answers is not in the document]
Comment (llun) [reply migrations/20260609101100_backfill_status_language.js]: Agreed. Removed the no-op migration in 4185af3 so it no longer clutters the migration tracker. The reason historical rows can't be backfilled (the persisted content blob keeps only the rendered fields, not the original `contentMap`) is now documented as a code comment on `getLanguage` in `lib/activities/note.ts`, and remains noted in the PR description.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by llun** [comment]
Thanks for the review. Addressed all three points in 4185af3:

- **`updateNoteJob` language coverage** — added two tests in `lib/jobs/updateNoteJob.test.ts`: one asserts an edit carrying `contentMap: { ja: '…' }` updates the stored `language` to `'ja'`, and one asserts an edit with no resolvable locale (`contentMap: {}`) preserves the existing `language` (`'th'`) instead of nulling it.
- **`firstLocaleKey` defensive guard** — added `typeof map !== 'object'`.
- **No-op migration** — removed; the backfill limitation is now a code comment on `getLanguage`.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by llun** [comment]
/gemini review

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by llun** [comment]
🤖 Sub-agent code review: Reviewed the federated status `language` changes (`lib/activities/note.ts`, `lib/jobs/createNoteJob.ts`, `lib/jobs/updateNoteJob.ts` + tests). No issues found.

Verified:
- `getLanguage` correctly guards against the array/Wordpress contentMap shape, `null`, and non-object primitives, and normalizes regional keys (e.g. `en-US` -> `en`) via `normalizeLanguageCode`.
- `createNoteJob` passes `string | null`, matching `CreateNoteParams.language?: string | null`.
- `updateNoteJob` maps `getLanguage(note) ?? undefined` so a missing locale becomes `undefined`, which `updateNote` (sql/status.ts L540: `language === undefined ? status.language : language`) treats as "keep existing". Behavior matches the documented intent and is covered by the preserve-on-empty test.
- No schema/migration impact (language lives in the JSON status content blob, read back at sql/status.ts L2440).
- Tests cover the meaningful edge cases; names are plain descriptive text and use table-driven `it.each` appropriately. No `fetch()` in components, no `console.*`. `lib/activities/note.test.ts` passes locally (36/36).

---
_Generated by [Claude Code](https://claude.ai/code)_

File: lib/activities/note.ts [hunk 3]
```diff
@@ -93,6 +94,35 @@ export const getContent = (object: BaseNote) => {
...
+ * This only works at ingestion time: the persisted status content blob keeps
+ * only the rendered fields and not the original `contentMap`, so the language of
+ * statuses federated before this helper existed cannot be recovered after the
+ * fact (they stay `language: null` until re-fetched or federated again).
+ */
+export const getLanguage = (object: BaseNote): string | null => {
+  const localeKey =
+    firstLocaleKey(object.contentMap) ?? firstLocaleKey(object.summaryMap)
+  if (!localeKey) return null
+  return normalizeLanguageCode(localeKey) || null
```
[the remark this answers is not in the document]
Comment (llun) [reply lib/activities/note.ts]: Good call. `getLanguage` now validates the normalized code against `/^[a-z]{2}$/` and returns `null` for malformed locale keys like `"12"`, `"!@"`, or `"a"` instead of persisting an invalid code. Added regression tests for the non-alphabetic and single-character cases. Fixed in f078fde.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by llun** [comment]
/gemini review

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by llun** [comment]
🤖 Sub-agent code review (round 3): Reviewed the current diff at f078fde across `lib/activities/note.ts`, `lib/jobs/createNoteJob.ts`, `lib/jobs/updateNoteJob.ts` and their tests.

No new issues found — the changed lines look correct and well covered:

- `getLanguage` + `firstLocaleKey` correctly handle the locale-keyed map, regional-key normalization (`en-US` → `en`), the array/Wordpress shape, empty maps, the `summaryMap` fallback, and the contentMap-over-summaryMap precedence. The `/^[a-z]{2}$/` guard after `normalizeLanguageCode` rejects malformed keys (`"12"`, `"a"`, `"!@"`).
- `updateNoteJob` uses `getLanguage(note) ?? undefined`, which matches `updateNote`'s documented semantics (`language === undefined` → keep existing, otherwise overwrite). The "refresh on edit with contentMap" and "preserve when edit has no contentMap" tests confirm both branches.
- `createNoteJob` persists the derived language with a sensible `null` default.

Earlier-round items (typeof guard, removed no-op migration, updateNoteJob language tests, getLanguage validation) are all present.

Minor non-blocking observation (not raising as a change request): `note.ts` is the first module in `lib/activities` to import from `lib/services` (`normalizeLanguageCode`). No circular dependency or runtime risk, but if strict layering matters, the pure ISO-639 normalizer could live in a lower-level util. Optional.

LGTM.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: lib/activities/note.ts [hunk 4]
```diff
@@ -93,6 +94,38 @@ export const getContent = (object: BaseNote) => {
...
+ */
+export const getLanguage = (object: BaseNote): string | null => {
+  const localeKey =
+    firstLocaleKey(object.contentMap) ?? firstLocaleKey(object.summaryMap)
+  if (!localeKey) return null
+  // Reject malformed locale keys (e.g. "12", "!@", "a") so only a valid
+  // two-letter ISO 639-1 code is persisted.
+  const language = normalizeLanguageCode(localeKey)
+  return /^[a-z]{2}$/.test(language) ? language : null
+}
```
[the remark this answers is not in the document]
Comment (llun) [reply lib/activities/note.ts:133]: Great catch — this was a real correctness bug. Fixed in d22f0d3: `getLanguage` now extracts the primary subtag (dropping any `-`/`_` regional suffix) and validates its length **before** normalizing, so a 3-letter ISO 639-2/3 code like `fil` or `ast` is rejected (→ `null`) instead of being truncated to the wrong 2-letter language. I kept `normalizeLanguageCode` for the final lower-casing of the already-validated 2-letter subtag (the task asked to reuse the shared normalizer rather than duplicate it), with the length check guarding against its `.slice(0,2)` truncation.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: lib/activities/note.ts [hunk 5]
```diff
@@ -1,3 +1,4 @@
+import { normalizeLanguageCode } from '@/lib/services/translation/types'
```
[the remark this answers is not in the document]
Comment (llun) [reply lib/activities/note.ts:1]: Keeping the `normalizeLanguageCode` import: rather than inlining the normalization, the fix validates the primary subtag length first and then still calls `normalizeLanguageCode` to lower-case the validated 2-letter code. This avoids the truncation bug while honoring the project's "reuse the shared normalizer, don't duplicate the logic" convention, so the import is still used (verified by `yarn lint` / `yarn build`).

---
_Generated by [Claude Code](https://claude.ai/code)_

File: lib/activities/note.test.ts [hunk 6]
```diff
@@ -299,4 +300,60 @@ describe('note entity utilities', () => {
...
+      {
+        description: 'returns null for a non-alphabetic locale key',
+        note: { type: 'Note', contentMap: { '12': '<p>Hello</p>' } },
+        expected: null
+      },
+      {
+        description: 'returns null for a single-character locale key',
+        note: { type: 'Note', contentMap: { a: '<p>Hello</p>' } },
+        expected: null
+      }
```
[the remark this answers is not in the document]
Comment (llun) [reply lib/activities/note.test.ts:365]: Added in d22f0d3 — a table case asserting `contentMap: { fil: '…' }` resolves to `null` (no incorrect truncation to `fi`), plus an `en_US` underscore-regional case that normalizes to `en`.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by llun** [comment]
/gemini review

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by llun** [comment]
🤖 Sub-agent code review (round 4): Reviewed the changed lines in `lib/activities/note.ts`, `lib/jobs/createNoteJob.ts`, `lib/jobs/updateNoteJob.ts` and their test files at d22f0d3. No new issues found — the diff is clean.

The new `getLanguage` primary-subtag logic handles the edge cases I checked:
- Leading/trailing separators (`"-en"` → empty primary subtag → rejected by `/^[a-z]{2}$/i`).
- Mixed/regional case (`"EN-us"`, `"en_US"` → normalized to `"en"`).
- Whitespace in keys (`.trim()` before the length check).
- Malformed keys (`"12"`, `"a"`) and 3-letter ISO 639-2/3 codes (`"fil"`) correctly return `null` rather than being truncated.
- Array/Wordpress `contentMap` and empty maps return `null`.

The `updateNoteJob` `?? undefined` "keep" semantics match `updateNote` in `lib/database/sql/status.ts` (`language === undefined ? status.language : language`), and the test coverage (refresh vs. preserve, plus the table-driven `getLanguage` cases) is thorough and follows repo conventions.

---
_Generated by [Claude Code](https://claude.ai/code)_
